lib/autoupdate.ps1

```diff
@@ -431,3 +424,104 @@ function autoupdate([String] $app, $dir, $json, [String] $version, [Hashtable] $
         Write-Host -f DarkGray "No updates for $app"
     }
 }
+
+## Helper Functions
+
+function PropertyHelper {
+    <#
+    .SYNOPSIS
+        Helper of updating property
+    .DESCRIPTION
+        Update manifest property (String, Array or PSCustomObject).
+    .PARAMETER Property
+        Property to be updated
+    .PARAMETER Value
+        New property values
+        Update line by line
+    .OUTPUTS
+        System.Object[]
+            The first element is new property, the second element is change flag
+    #>
+    param (
+        [Object]$Property,
+        [Object]$Value
+    )
+    $hasChanged = $false
+    switch ($Property.GetType().Name) {
+        'String' {
+            $Value = $Value -as [String]
+            if (($null -ne $Value) -and ($value -ne $Property)) {
+                $Property = $Value
+                $hasChanged = $true
+            }
+        }
+        'Object[]' {
+            $Value = @($Value)
+            for ($i = 0; $i -lt [Math]::Min($Property.Length, $Value.Length); $i++) {
+                $Property[$i], $hasItemChanged = PropertyHelper -Property $Property[$i] -Value $Value[$i]
+                $hasChanged = $hasChanged -or $hasItemChanged
+            }
+        }
+        'PSCustomObject' {
+            if ($Value -is [PSObject]) {
+                foreach ($name in $Property.PSObject.Properties.Name) {
+                    if ($Value.$name) {
+                        $Property.$name, $hasItemChanged = PropertyHelper -Property $Property.$name -Value $Value.$name
+                        $hasChanged = $hasChanged -or $hasItemChanged
+                    }
+                }
+            }
+        }
+    }
+    return $Property, $hasChanged
+}
+
+function HashHelper {
+    <#
+    .SYNOPSIS
+        Helper of getting file hash(es)
+    .DESCRIPTION
+        Get file hash(es) by hash extraction template(s).
```

Comment:
By `hash extraction templates` you mean hash property in autoupdate right?

Reply:
Yes, what's your advice for this property? Or just `autoupdate.hash`?

Reply:
What about:

```suggestion
        Get file hash(es) by hash extraction method.
```

```suggestion
        Extract or calculate file hash(es)
```